src/esrc.rs

```diff
@@ -0,0 +1,195 @@
+    }
+}
+
+#[derive(Copy, Clone, PartialEq, Eq, Debug)]
+struct DecryptionKey(Scalar);
+impl Mul<&DecryptionKey> for &G1Projective {
+    type Output = G1Projective;
+
+    fn mul(self, rhs: &DecryptionKey) -> Self::Output {
+        self * rhs.0
+    }
+}
+
+#[derive(Copy, Clone, Debug)]
+struct CipherText(G1Projective, G1Projective);
+
+#[derive(Clone, Copy, PartialEq, Eq, Debug)]
+struct SigningKey(Scalar, Scalar);
+#[derive(Copy, Clone, Debug)]
+struct VerifyKey(G2Projective, G2Projective);
+#[derive(Copy, Clone, Debug)]
+struct Signature(G1Projective, G1Projective, G2Projective, G1Projective);
+
+#[allow(dead_code)]
+fn setup_encryption_keys(rng: &mut (impl RngCore + CryptoRng)) -> (EncryptionKey, DecryptionKey) {
+    let dec_key = Scalar::from(rng.next_u64());
+    let enc_key = G1Affine::generator() * dec_key;
+    (EncryptionKey(enc_key), DecryptionKey(dec_key))
+}
+
+#[allow(dead_code)]
+fn setup_signing_keys(rng: &mut (impl RngCore + CryptoRng)) -> (SigningKey, VerifyKey) {
+    let x0 = Scalar::from(rng.next_u64()).double();
+    let x1 = Scalar::from(rng.next_u64()).double();
+    let x_hat0 = G2Affine::generator() * x0;
+    let x_hat1 = G2Affine::generator() * x1;
+    (SigningKey(x0, x1), VerifyKey(x_hat0, x_hat1))
+}
+
+#[allow(dead_code)]
+fn encrypt(
+    rng: &mut (impl RngCore + CryptoRng),
+    m: G1Affine,
+    enc_key: &EncryptionKey,
+) -> CipherText {
+    let r = Scalar::from(rng.next_u64());
+    let c0 = G1Affine::generator() * r;
+    let c1 = m + (enc_key * r);
+    CipherText(c0, c1)
+}
+
+#[allow(dead_code)]
+fn decrypt(cm: &CipherText, dec_key: &DecryptionKey) -> G1Affine {
+    let CipherText(c0, c1) = cm;
+    let dec = c1 - (c0 * dec_key);
+    G1Affine::from(dec)
+}
+
+#[allow(dead_code)]
+fn randomize(cm: &CipherText, enc_key: &EncryptionKey, r_prime: Scalar) -> CipherText {
+    let CipherText(c0, c1) = cm;
+    let rc0 = c0 + (G1Affine::generator() * r_prime);
+    let rc1 = c1 + (enc_key * r_prime);
+    CipherText(rc0, rc1)
+}
+
+#[allow(dead_code)]
+fn sign(
+    rng: &mut (impl RngCore + CryptoRng),
+    cm: &CipherText,
+    enc_key: &EncryptionKey,
+    sign_key: SigningKey,
+) -> Signature {
+    let scalar_s = Scalar::from(rng.next_u64());
+    let scalar_s_inverted = scalar_s.invert().unwrap(); // TODO: how to handle?
+    let CipherText(c0, c1) = cm;
+    let SigningKey(x0, x1) = sign_key;
+    let z = (G1Affine::generator() + (c0 * x0) + (c1 * x1)) * scalar_s_inverted;
+    let s = G1Affine::generator() * scalar_s;
+    let s_hat = G2Affine::generator() * scalar_s;
+    let t = ((G1Affine::generator() * x0) + (enc_key * x1)) * scalar_s_inverted;
+    Signature(z, s, s_hat, t)
+}
+
+#[allow(dead_code)]
+fn adapt(
+    rng: &mut (impl RngCore + CryptoRng),
+    signature: &Signature,
+    r_prime: Scalar,
+) -> Signature {
+    let scalar_s_prime = Scalar::from(rng.next_u64());
+    let scalar_s_prime_inverted = scalar_s_prime.invert().unwrap(); // TODO: how to handle?
```

Comment:
My understanding is that it should be incredibly rare for this to be non-invertible, so I think it makes sense to handle this in this function, since it's basically just a need to retry. I think we could cap the retries at a reasonable level (say 10 or even 100) so that it does fail if there happens to be a bug in the provided `rng`.

Reply:
done